// odds-api-io/dist/cjs/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OddsAPIClient = void 0;
const errors_js_1 = require("./errors.js");
const DEFAULT_BASE_URL = 'https://api2.odds-api.io/v3';
const DEFAULT_TIMEOUT = 10000;
/**
 * Official Node.js client for Odds-API.io
 *
 * @example
 * ```typescript
 * import { OddsAPIClient } from 'odds-api-io';
 *
 * const client = new OddsAPIClient({ apiKey: 'your-api-key' });
 * const sports = await client.getSports();
 * ```
 */
class OddsAPIClient {
    constructor(config) {
        this.apiKey = config.apiKey;
        this.baseUrl = config.baseUrl || DEFAULT_BASE_URL;
        this.timeout = config.timeout || DEFAULT_TIMEOUT;
    }
    /**
     * Make a GET request to the API
     */
    async request(path, params = {}, requiresAuth = false) {
        const url = new URL(`${this.baseUrl}/${path}`);
        // Add query parameters
        if (requiresAuth) {
            params.apiKey = this.apiKey;
        }
        Object.entries(params).forEach(([key, value]) => {
            if (value !== undefined && value !== null) {
                url.searchParams.append(key, String(value));
            }
        });
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeout);
        try {
            const response = await fetch(url.toString(), {
                method: 'GET',
                signal: controller.signal,
                headers: {
                    'User-Agent': 'odds-api-io-node-sdk/1.0.0',
                },
            });
            clearTimeout(timeoutId);
            if (!response.ok) {
                await this.handleErrorResponse(response);
            }
            return (await response.json());
        }
        catch (error) {
            clearTimeout(timeoutId);
            if (error.name === 'AbortError') {
                throw new errors_js_1.TimeoutError(`Request timeout after ${this.timeout}ms`);
            }
            if (error instanceof errors_js_1.OddsAPIError) {
                throw error;
            }
            throw new errors_js_1.NetworkError(`Network request failed: ${error.message}`);
        }
    }
    /**
     * Make a PUT request to the API
     */
    async requestPut(path, params = {}, requiresAuth = false) {
        const url = new URL(`${this.baseUrl}/${path}`);
        // Add query parameters
        if (requiresAuth) {
            params.apiKey = this.apiKey;
        }
        Object.entries(params).forEach(([key, value]) => {
            if (value !== undefined && value !== null) {
                url.searchParams.append(key, String(value));
            }
        });
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeout);
        try {
            const response = await fetch(url.toString(), {
                method: 'PUT',
                signal: controller.signal,
                headers: {
                    'User-Agent': 'odds-api-io-node-sdk/1.0.0',
                },
            });
            clearTimeout(timeoutId);
            if (!response.ok) {
                await this.handleErrorResponse(response);
            }
            return (await response.json());
        }
        catch (error) {
            clearTimeout(timeoutId);
            if (error.name === 'AbortError') {
                throw new errors_js_1.TimeoutError(`Request timeout after ${this.timeout}ms`);
            }
            if (error instanceof errors_js_1.OddsAPIError) {
                throw error;
            }
            throw new errors_js_1.NetworkError(`Network request failed: ${error.message}`);
        }
    }
    /**
     * Handle error responses from the API
     */
    async handleErrorResponse(response) {
        const status = response.status;
        let errorMessage;
        try {
            const errorBody = await response.text();
            errorMessage = errorBody || response.statusText;
        }
        catch {
            errorMessage = response.statusText;
        }
        switch (status) {
            case 401:
                throw new errors_js_1.InvalidAPIKeyError('Invalid API key');
            case 429:
                throw new errors_js_1.RateLimitExceededError('Rate limit exceeded - please wait before retrying');
            case 404:
                throw new errors_js_1.NotFoundError('Resource not found');
            default:
                throw new errors_js_1.OddsAPIError(`API error ${status}: ${errorMessage}`);
        }
    }
    // ============================================================================
    // SPORTS & LEAGUES
    // ============================================================================
    /**
     * Get all available sports
     *
     * @returns List of available sports
     *
     * @example
     * ```typescript
     * const sports = await client.getSports();
     * console.log(`Found ${sports.length} sports`);
     * ```
     */
    async getSports() {
        return this.request('sports');
    }
    /**
     * Get leagues for a specific sport
     *
     * @param sport - Sport identifier (e.g., 'basketball', 'football')
     * @returns List of leagues
     *
     * @example
     * ```typescript
     * const leagues = await client.getLeagues('basketball');
     * ```
     */
    async getLeagues(sport) {
        return this.request('leagues', { sport }, true);
    }
    // ============================================================================
    // EVENTS
    // ============================================================================
    /**
     * Get events with optional filters
     *
     * @param params - Event filter parameters
     * @returns List of events
     *
     * @example
     * ```typescript
     * const events = await client.getEvents({
     *   sport: 'basketball',
     *   league: 'usa-nba'
     * });
     * ```
     */
    async getEvents(params) {
        return this.request('events', params, true);
    }
    /**
     * Get a specific event by ID
     *
     * @param eventId - Event ID
     * @returns Event details
     *
     * @example
     * ```typescript
     * const event = await client.getEventById('62924717');
     * ```
     */
    async getEventById(eventId) {
        return this.request(`events/${eventId}`, {}, true);
    }
    /**
     * Get currently live events for a sport
     *
     * @param sport - Sport identifier
     * @returns List of live events
     *
     * @example
     * ```typescript
     * const liveEvents = await client.getLiveEvents('basketball');
     * ```
     */
    async getLiveEvents(sport) {
        return this.request('events/live', { sport }, true);
    }
    /**
     * Search for events by keyword
     *
     * @param query - Search query
     * @returns List of matching events
     *
     * @example
     * ```typescript
     * const events = await client.searchEvents('Lakers');
     * ```
     */
    async searchEvents(query) {
        return this.request('events/search', { query }, true);
    }
    // ============================================================================
    // HISTORICAL
    // ============================================================================
    /**
     * Get finished historical events within a date range
     *
     * @param params - Historical event filter parameters
     * @returns List of historical events
     *
     * @example
     * ```typescript
     * const events = await client.getHistoricalEvents({
     *   sport: 'football',
     *   league: 'england-premier-league',
     *   from: '2026-01-01T00:00:00Z',
     *   to: '2026-01-31T23:59:59Z'
     * });
     * ```
     */
    async getHistoricalEvents(params) {
        return this.request('historical/events', params, true);
    }
    /**
     * Get closing odds for a historical event
     *
     * @param params - Historical odds query parameters
     * @returns Historical event odds data
     *
     * @example
     * ```typescript
     * const odds = await client.getHistoricalOdds({
     *   eventId: '67697888',
     *   bookmakers: 'singbet,bet365'
     * });
     * ```
     */
    async getHistoricalOdds(params) {
        return this.request('historical/odds', params, true);
    }
    // ============================================================================
    // ODDS
    // ============================================================================
    /**
     * Get odds for a specific event
     *
     * @param params - Odds query parameters
     * @returns Event odds data
     *
     * @example
     * ```typescript
     * const odds = await client.getEventOdds({
     *   eventId: '62924717',
     *   bookmakers: 'singbet,bet365'
     * });
     * ```
     */
    async getEventOdds(params) {
        return this.request('odds', params, true);
    }
    /**
     * Track odds movements for an event
     *
     * @param params - Odds movement parameters
     * @returns Historical odds movements
     *
     * @example
     * ```typescript
     * const movements = await client.getOddsMovement({
     *   eventId: '62924717',
     *   bookmaker: 'singbet',
     *   market: 'moneyline'
     * });
     * ```
     */
    async getOddsMovement(params) {
        return this.request('odds/movements', params, true);
    }
    /**
     * Get odds for multiple events at once
     *
     * @param params - Multi-event odds parameters
     * @returns Array of event odds
     *
     * @example
     * ```typescript
     * const odds = await client.getOddsForMultipleEvents({
     *   eventIds: '12345,67890',
     *   bookmakers: 'singbet,bet365'
     * });
     * ```
     */
    async getOddsForMultipleEvents(params) {
        return this.request('odds/multi', params, true);
    }
    /**
     * Get odds updated since a given timestamp
     *
     * @param params - Updated odds parameters
     * @returns Array of updated event odds
     *
     * @example
     * ```typescript
     * const updatedOdds = await client.getUpdatedOddsSince({
     *   since: Date.now() - 3600000, // Last hour
     *   bookmaker: 'singbet',
     *   sport: 'basketball'
     * });
     * ```
     */
    async getUpdatedOddsSince(params) {
        return this.request('odds/updated', params, true);
    }
    // ============================================================================
    // PARTICIPANTS
    // ============================================================================
    /**
     * Get participants (teams/players) for a sport
     *
     * @param params - Participant query parameters
     * @returns List of participants
     *
     * @example
     * ```typescript
     * const participants = await client.getParticipants({
     *   sport: 'basketball',
     *   search: 'Warriors'
     * });
     * ```
     */
    async getParticipants(params) {
        return this.request('participants', params, true);
    }
    /**
     * Get a specific participant by ID
     *
     * @param participantId - Participant ID
     * @returns Participant details
     *
     * @example
     * ```typescript
     * const participant = await client.getParticipantById(3428);
     * ```
     */
    async getParticipantById(participantId) {
        return this.request(`participants/${participantId}`, {}, true);
    }
    // ============================================================================
    // BOOKMAKERS
    // ============================================================================
    /**
     * Get all available bookmakers
     *
     * @returns List of bookmakers
     *
     * @example
     * ```typescript
     * const bookmakers = await client.getBookmakers();
     * ```
     */
    async getBookmakers() {
        return this.request('bookmakers');
    }
    /**
     * Get your selected bookmakers
     *
     * @returns List of selected bookmakers
     *
     * @example
     * ```typescript
     * const selected = await client.getSelectedBookmakers();
     * ```
     */
    async getSelectedBookmakers() {
        return this.request('bookmakers/selected', {}, true);
    }
    /**
     * Select specific bookmakers for your account
     *
     * @param bookmakers - Comma-separated bookmaker IDs
     * @returns Success response
     *
     * @example
     * ```typescript
     * await client.selectBookmakers('singbet,bet365');
     * ```
     */
    async selectBookmakers(bookmakers) {
        return this.requestPut('bookmakers/selected/select', { bookmakers }, true);
    }
    /**
     * Clear your selected bookmakers
     *
     * @returns Success response
     *
     * @example
     * ```typescript
     * await client.clearSelectedBookmakers();
     * ```
     */
    async clearSelectedBookmakers() {
        return this.requestPut('bookmakers/selected/clear', {}, true);
    }
    // ============================================================================
    // BETTING ANALYSIS
    // ============================================================================
    /**
     * Find arbitrage betting opportunities
     *
     * @param params - Arbitrage bet parameters
     * @returns List of arbitrage opportunities
     *
     * @example
     * ```typescript
     * const arbs = await client.getArbitrageBets({
     *   bookmakers: 'singbet,bet365',
     *   limit: 10,
     *   includeEventDetails: true
     * });
     * ```
     */
    async getArbitrageBets(params) {
        return this.request('arbitrage-bets', params, true);
    }
    /**
     * Find value betting opportunities
     *
     * @param params - Value bet parameters
     * @returns List of value bets
     *
     * @example
     * ```typescript
     * const valueBets = await client.getValueBets({
     *   bookmaker: 'singbet',
     *   includeEventDetails: true
     * });
     * ```
     */
    async getValueBets(params) {
        return this.request('value-bets', params, true);
    }
}
exports.OddsAPIClient = OddsAPIClient;
