// odds-api-io/dist/cjs/errors.js
"use strict";
/**
 * Custom error classes for Odds-API.io client
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkError = exports.TimeoutError = exports.NotFoundError = exports.RateLimitExceededError = exports.InvalidAPIKeyError = exports.OddsAPIError = void 0;
/**
 * Base error class for all Odds-API.io errors
 */
class OddsAPIError extends Error {
    constructor(message) {
        super(message);
        this.name = 'OddsAPIError';
        Object.setPrototypeOf(this, OddsAPIError.prototype);
    }
}
exports.OddsAPIError = OddsAPIError;
/**
 * Thrown when the API key is invalid or missing
 */
class InvalidAPIKeyError extends OddsAPIError {
    constructor(message = 'Invalid API key') {
        super(message);
        this.name = 'InvalidAPIKeyError';
        Object.setPrototypeOf(this, InvalidAPIKeyError.prototype);
    }
}
exports.InvalidAPIKeyError = InvalidAPIKeyError;
/**
 * Thrown when rate limit is exceeded
 */
class RateLimitExceededError extends OddsAPIError {
    constructor(message = 'Rate limit exceeded - please wait before retrying') {
        super(message);
        this.name = 'RateLimitExceededError';
        Object.setPrototypeOf(this, RateLimitExceededError.prototype);
    }
}
exports.RateLimitExceededError = RateLimitExceededError;
/**
 * Thrown when a resource is not found (404)
 */
class NotFoundError extends OddsAPIError {
    constructor(message = 'Resource not found') {
        super(message);
        this.name = 'NotFoundError';
        Object.setPrototypeOf(this, NotFoundError.prototype);
    }
}
exports.NotFoundError = NotFoundError;
/**
 * Thrown when request times out
 */
class TimeoutError extends OddsAPIError {
    constructor(message = 'Request timeout') {
        super(message);
        this.name = 'TimeoutError';
        Object.setPrototypeOf(this, TimeoutError.prototype);
    }
}
exports.TimeoutError = TimeoutError;
/**
 * Thrown when network request fails
 */
class NetworkError extends OddsAPIError {
    constructor(message) {
        super(message);
        this.name = 'NetworkError';
        Object.setPrototypeOf(this, NetworkError.prototype);
    }
}
exports.NetworkError = NetworkError;
