// odds-api-io/dist/cjs/index.js
"use strict";
/**
 * Odds-API.io Node.js SDK
 *
 * Official Node.js client for Odds-API.io - Real-time sports betting odds from 250+ bookmakers
 *
 * @packageDocumentation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkError = exports.TimeoutError = exports.NotFoundError = exports.RateLimitExceededError = exports.InvalidAPIKeyError = exports.OddsAPIError = exports.OddsAPIClient = void 0;
var client_js_1 = require("./client.js");
Object.defineProperty(exports, "OddsAPIClient", { enumerable: true, get: function () { return client_js_1.OddsAPIClient; } });
var errors_js_1 = require("./errors.js");
Object.defineProperty(exports, "OddsAPIError", { enumerable: true, get: function () { return errors_js_1.OddsAPIError; } });
Object.defineProperty(exports, "InvalidAPIKeyError", { enumerable: true, get: function () { return errors_js_1.InvalidAPIKeyError; } });
Object.defineProperty(exports, "RateLimitExceededError", { enumerable: true, get: function () { return errors_js_1.RateLimitExceededError; } });
Object.defineProperty(exports, "NotFoundError", { enumerable: true, get: function () { return errors_js_1.NotFoundError; } });
Object.defineProperty(exports, "TimeoutError", { enumerable: true, get: function () { return errors_js_1.TimeoutError; } });
Object.defineProperty(exports, "NetworkError", { enumerable: true, get: function () { return errors_js_1.NetworkError; } });
